src/rules/no-incompatible-branch-types.coffee

```diff
@@ -0,0 +1,63 @@
+ts = require 'typescript'
+
+# getSourceFileOfNode = (tsNode) ->
+#   while node and node.kind isnt ts.SyntaxKind.SourceFile
+#     node = node.parent
+#   node
+
+module.exports =
+  meta:
+    docs:
+      description:
+        'Flag incompatible early-return types in branch()/branchPure()'
+      category: 'Possible Errors'
+      recommended: yes
+    schema: []
+
+  create: (context) ->
+    {parserServices} = context
+    return {} unless parserServices?.hasFullTypeInformation
+
+    typeChecker = parserServices.program.getTypeChecker()
+
+    CallExpression: (node) ->
+      {callee, parent} = node
+      return unless (
+        callee?.type is 'Identifier' and
+        callee.name in [### 'renderNothing', ### 'returns']
+      )
+
+      return unless (
+        parent.type is 'CallExpression' and
+        parent.callee.type is 'Identifier' and
+        parent.callee.name in ['branch' ### , 'branchPure'###]
+      )
+
+      {arguments: [returnsCallbackNode]} = node
+      return unless returnsCallbackNode?
+
+      tsReturnsCallbackNode = parserServices.esTreeNodeToTSNodeMap.get(
+        returnsCallbackNode
+      )
+      tsReturnsCallbackType = typeChecker.getTypeAtLocation(
+        tsReturnsCallbackNode
+      )
+      [signature] = typeChecker.getSignaturesOfType(
+        tsReturnsCallbackType
+        ts.SignatureKind.Call
+      )
+      return unless signature?
+      returnType = typeChecker.getReturnTypeOfSignature signature
+      returnTypeString = typeChecker.typeToString returnType
+      return if returnTypeString in ['null', 'Element', 'Element | null']
+
+      # tsProgram = parserServices.program
+      # sourceFile = getSourceFileOfNode tsReturnsCallbackNode
+      # diagnostics = tsProgram.getSemanticDiagnostics sourceFile
```

Comment:
I pulled this from a `typescript-eslint` rule, was helpful when debugging to see what Typescript was reporting as the type errors ("semantic diagnostics") for the test files